### PR TITLE
fix Issue 21301 - Wrong values being passed in long parameter list

### PR DESCRIPTION
**Pull request by WalterBright**
What a terrible bug this is.

### CODE REVIEW

**Comment by WalterBright** [comment]
It's definitely an alignment problem. The caller is aligning p3 to 16 bytes, the caller is aligning it to 8.

**Comment by WalterBright** [comment]
@RazvanN7 is it true that the compiler has to be able to build a working version of itself before it runs the test suite?

**Comment by RazvanN7** [comment]
@WalterBright I'm not sure what you are referring to, but we have the following situations:

- cicleci uses uses a stable version of the compiler binary (let's call it version A) to build the new version of the compiler (let's call it version B). Then it uses version B to build version C of the compiler from the committed sources. It then uses version C to run the tests.

- all the other tests use version B to run the tests.

**Comment by dkorpel** [comment]
Cirrus fails currently because it cannot build the test runner (d_do_test.d) with the new compiler (version B).

**Comment by WalterBright** [comment]
The test suite is designed to test the compiler. The most effective way to do this is:

1. build the test runner with the previous working version of the compiler
2. build checkwhitespace with the previous working version of the compiler
3. build the library with the previous working version of the compiler
4. build the new compiler
5. run the test suite
6. run the library unit tests
7. now do 1..6 with the new build of the compiler

Instead, what currently appears to happens is:

1. build the new compiler with the previous working version of the compiler
2. new compiler crashes building the test runner
3. new compiler crashes building checkwhitespace
4. new compiler crashes building library
5. test runner built with new compiler crashes
8. checkwhitespace built with new compiler crashes

This makes for much harder debugging what went wrong.

I recommend that at least one of the tests, like maybe the autotester, run the test suite *before* attempting to compile anything else with the new compiler. The test suite is already filled with minimized test cases.

**Comment by Herringway** [comment]
> 1. build the new compiler with the previous working version of the compiler
> 2. new compiler crashes building the test runner
> 3. new compiler crashes building checkwhitespace
> 4. new compiler crashes building library
> 5. test runner built with new compiler crashes
> 6. checkwhitespace built with new compiler crashes

How do the test runner and checkwhitespace end up being built if the compiler crashed while building them?

**Comment by WalterBright** [comment]
They don't end up being built. That's the problem.

**Comment by WalterBright** [comment]
The trouble here is the type of the float[3] function argument was set to TYcdouble to make handling it more convenient in in the backend. But the alignment of TYcdouble was set to 16, which is incorrect. It should be 8. The misalignment caused pushing the TYcdouble argument on the stack to include an extra 8 bytes, while the float[3] parameter was aligned on 8 bytes.

File: compiler/src/dmd/backend/dtype.d [hunk 1]
```diff
@@ -304,6 +304,10 @@ L1:
             case TYldouble:
                 assert(0);
 
+          case TYcdouble:
```
Comment (nordlow) [inline compiler/src/dmd/backend/dtype.d:307]: Indentation is off. Adjust two characters to the right.

File: compiler/src/dmd/backend/dtype.d [hunk 2]
```diff
@@ -304,6 +304,10 @@ L1:
             case TYldouble:
                 assert(0);
 
+            case TYcdouble:
+              sz = 8;         // not 16
+              break;
```
Comment (nordlow) [inline compiler/src/dmd/backend/dtype.d:309]: These two lines also needs indentation two characters to the right.

**Review by RazvanN7 — APPROVED** [review]

**Comment by ibuclaw** [comment]
Regressions should definitely target stable.

**Pull request merged**